subsys/net/lib/azure_fota/azure_fota.c

```diff
@@ -175,21 +266,98 @@ static void fota_dl_handler(const struct fota_download_evt *fota_evt)
+				    const char *current_fw,
+				    const char *pending_fw)
+{
+	bool report_needed;
+
+	switch (reported_status_get(status)) {
+	case REP_STATUS_CURRENT:
+		/* "current" status indicates no FOTA was in progress, check if the
+		 * reported version matches the actual version of running firmware.
+		 */
+		if (strcmp(get_current_version(), current_fw) == 0) {
+			LOG_DBG("Currently reported version is correct");
+			rep_status_set(REP_STATUS_CURRENT);
+			report_needed = false;
+		} else {
+			/* Reported current version should be updated. */
+			LOG_DBG("Currently reported version must be updated");
+			rep_status_set(REP_STATUS_CURRENT);
+			report_needed = true;
+		}
+
+		break;
+	case REP_STATUS_APPLYING:
+		/* "applying" indicates firmware was downloaded and would be
+		 * used after reboot. Check if running version matches the
+		 * pending version.
+		 */
+		report_needed = true;
+
+		if (strcmp(get_current_version(), pending_fw) == 0) {
+			LOG_DBG("FOTA firmware was successfully applied");
+			rep_status_set(REP_STATUS_CURRENT);
+		} else if (strcmp(get_current_version(), current_fw) == 0) {
+			LOG_WRN("FOTA image was not applied, rolled back");
+			rep_status_set(REP_STATUS_ROLLEDBACK);
+		} else {
+			/* Current version matches neither of the reported
+			 * current or pending versions, which indicate that
+			 * the firmware has been changed by other means.
+			 */
+			rep_status_set(REP_STATUS_CURRENT);
+		}
+
+		break;
+	case REP_STATUS_DOWNLOADING:
+		/* A download was in progress and must have failed for this
+		 * situation to occur. Therefore, the error must be reported.
+		 */
+		report_needed = true;
+
+		rep_status_set(REP_STATUS_ERROR);
+
+		break;
+	default:
+		/* In all other cases, the current firmware should be reported
+		 * back, and the status is "current".
+		 */
+		report_needed = true;
+
+		rep_status_set(REP_STATUS_CURRENT);
+
+		break;
+	}
+
+	return report_needed;
+}
+
 /* Checks the 'reported' object for current FOTA state
- * Returns true if the fota_state has been updated and should be reported to
+ * Returns true if the current_status has been updated and should be reported to
  * the device twin. If no action is needed, false is returned.
  */
 static bool parse_reported_status(const char *msg)
 {
 	struct cJSON *root_obj, *reported_obj, *fw_obj, *fw_status_obj,
-		*fw_current_obj, *fw_pending_obj;
+		*fw_current_obj, *fw_pending_obj, *job_id_obj;
 	bool report_needed = false;
 
-	/* The 'reported' object has the excpected structure if it exists:
+	/* The 'reported' object has the expected structure if it exists:
 	 *	properties.reported.firmware : {
-			"fwUpdateStatus": <status>,
-			"currentFwVersion": <current firmware version>,
-			"pendingFwVersion": <FOTA version in progress (if any)>
+	 *		"fwUpdateStatus": <status>,
+	 *		"currentFwVersion": <current firmware version>,
+	 *		"pendingFwVersion": <FOTA version in progress (if any)>,
+	 *		"jobId": <FOTA update ID, formatted according to
```

Comment:
Does the ```AZURE_FOTA``` actually check if the jobID is in a format according to RFC4122? If not, it may be a little misleading to write that it is expected.

Reply:
It doesn't, so that reference is mainly to make sure users don't have larger strings than 32 hex chars + 4x `-`. Can be changed if desired